Handle mixed scalar-types in Eigen for symbolic classes

soonho-tri opened the pull request:
Multiplications among Eigen matrices of `{symbolic::Expression, symbolic::Variable, double}` were not properly handled previously. If operands have dynamic sizes, it causes compile-time errors.

According to [1], this is an issue on Eigen side. This patch is a workaround explicitly providing operator overloading among them.

[1]: http://stackoverflow.com/questions/41494288/mixing-scalar-types-in-eigen

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4702)
<!-- Reviewable:end -->

soonho-tri commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/symbolic_expression.cc, line 417 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYO5r_-eMMM87klO_:-K_pYO5r_-eMMM87klOa:b-3xhsfx) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_expression.cc#L417)):*
> ```cpp
> Expression operator-(Expression lhs, const Variable& rhs) { return lhs -= rhs; }
> Expression operator-(const Variable& lhs, const Expression& rhs) {
>   return Expression(lhs) - rhs;
> ```

Previously, we had `var_x - expr_x = -0.0` instead of `0.0`. This change fixes that issue. 

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_pYeylg5hvJnyMJjoL:bs7ryss) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
> ```objc
> 
>   /** Conversion to bool. */
>   explicit operator bool() const { return Evaluate(); }
> ```

This is required by Eigen to handle `MatrixX<Expression>` * `VectorX<Expression>`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
+@hongkai-dai , could you do feature-review please?
+@jwnimmer-tri , could you do platform-review please?

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pZ8rPhEFtnBQCSIis:-K_pZ8rPhEFtnBQCSIit:b-6ua79) (no related file):*
@hongkai-dai , as soon as this one closes, I'll open a PR introducing the symbolic version of `AddLinearConstraint` method. I need this one to test that PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-:-K_pYx4BKHlIzywdymhK:byuwv5d)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pgxcikjTJ2-XpjP8Z:-K_pgxcikjTJ2-XpjP8_:b-wkexd6) (no related file):*
@soonho-tri Please don't "mark files as reviewed" when you are the author; it screws up the comment discussion workflow.

---

*[drake/common/symbolic_expression.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pgcIk7ET-roe39QeG:-K_pgcIlJVh8psHeJSRV:b8lctiz) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_expression.h#L17)):*
> ```objc
> #include "drake/common/cond.h"
> #include "drake/common/dummy_value.h"
> #include "drake/common/eigen_types.h"
> ```

BTW Does this header actually use any of the abbreviations in eigen_types?

---

*[drake/common/symbolic_expression.h, line 505 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pgBbxFbb6u_H0cqGl:-K_pgBbyXH0mCGzmPzIR:b-g6su0o) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_expression.h#L505)):*
> ```objc
>     Eigen::Matrix<Expression, MatrixL::RowsAtCompileTime,
>                   MatrixR::ColsAtCompileTime> >::type
> operator*(MatrixL const& lhs, MatrixR const& rhs) {
> ```

BTW *Meta* should these consts be formatted according to our local style?

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_phGqiKCY7hyP5u6Ia:bokbtqp) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

This is required by Eigen to handle `MatrixX<Expression>` * `VectorX<Expression>`.
</blockquote></details>

Please provide more detail (eventually maybe in a source comment, but reviewable discussion is maybe better until we settle this).  This seems undesirable.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_pfc1wQoLkJbkt51Ht:bqy8cmq) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
> ```cpp
>   VectorX<Variable> V_var_dyn{2};
> 
>   Matrix2<double> M_double_fixed;
> ```

BTW I think putting this into a member field without alignment goop in the class violates Eigen's alignment contract.  Do we care?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pgxcikjTJ2-XpjP8Z:-K_ptJM_t2Xht4nWI1Ba:b-umt4h4) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@soonho-tri Please don't "mark files as reviewed" when you are the author; it screws up the comment discussion workflow.
</blockquote></details>

I see. I've formed a habit of doing this.. won't do it anymore.

---

*[drake/common/symbolic_expression.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pgcIk7ET-roe39QeG:-K_puED0dxANDGpzWS5X:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_expression.h#L17)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Does this header actually use any of the abbreviations in eigen_types?
</blockquote></details>

Done.

---

*[drake/common/symbolic_expression.h, line 505 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pgBbxFbb6u_H0cqGl:-K_puCMe_277A6fspOnI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_expression.h#L505)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW *Meta* should these consts be formatted according to our local style?
</blockquote></details>

Done.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_pw0afO1ZYF1miQtt9:bnz1e0w) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please provide more detail (eventually maybe in a source comment, but reviewable discussion is maybe better until we settle this).  This seems undesirable.
</blockquote></details>

Without it, we have the following build error:

```
external/eigen/Eigen/src/Core/GeneralProduct.h:250:58: error: value of type 'drake::symbolic::Formula' is not contextually convertible to 'bool'
      const bool alphaIsCompatible = (!ComplexByReal) || (numext::imag(actualAlpha)==RealScalar(0));
                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_pxkoUtlj_qKTZ1KPm:b-9owjom) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I think putting this into a member field without alignment goop in the class violates Eigen's alignment contract.  Do we care?
</blockquote></details>

Do you think we having `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` is a good idea here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_pyoiAcVzFRD7e24dl:b6yj39o) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Do you think we having `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` is a good idea here?
</blockquote></details>

My personal preference on Eigen alignment is to keep the fix encapsulated, so either declaring `::DontAlign` in the template argument of these member fields, or stashing them in a unique_ptr (possibly with reference alias for convenience).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 2052 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pzQuRFP4pwt9-GmJv:-K_pzQuStcMw9ct5jaKS:b3k5tuv) ([raw file](https://github.com/robotlocomotion/drake/blob/22d4fd878e5e3d254a7bb9bc48db5fbd0e40ca38/drake/common/test/symbolic_expression_test.cc#L2052)):*
> ```cpp
>   const Expression w_{var_w_};
> 
>   Matrix2<Expression> M_expr_fixed;
> ```

*Meta* missing trailing `_` on member fields.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_pzKsNz5FPM5wUKSQn:b-bcl9n3) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

My personal preference on Eigen alignment is to keep the fix encapsulated, so either declaring `::DontAlign` in the template argument of these member fields, or stashing them in a unique_ptr (possibly with reference alias for convenience).
</blockquote></details>

Is it the case that we have to worry about this only if one write `new SymbolicMixingScalarTypesTest ` in the future? If that's the case, I think it's safe to assume that it's not going to happen since it's a part of testcase.

Is GTEST doing that internally?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 2052 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pzQuRFP4pwt9-GmJv:-K_q-UQyIgE8WY8-_BGY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22d4fd878e5e3d254a7bb9bc48db5fbd0e40ca38/drake/common/test/symbolic_expression_test.cc#L2052)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

*Meta* missing trailing `_` on member fields.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/symbolic_expression.h, line 495 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_q0F1uAHQUoWXX5nQ2:-K_q0F1uAHQUoWXX5nQ3:b-xivduy) ([raw file](https://github.com/robotlocomotion/drake/blob/22d4fd878e5e3d254a7bb9bc48db5fbd0e40ca38/drake/common/symbolic_expression.h#L495)):*
> ```objc
>     const Expression& e);
> 
> // MatrixBase<Expression> * Matrix<Variable> => Matrix<Expression>
> ```

`Matrix<Expression>` instead of `MatrixBase<Expression>`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_q10Le0CIBNY-30Pii:by5sz3n) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Is it the case that we have to worry about this only if one write `new SymbolicMixingScalarTypesTest ` in the future? If that's the case, I think it's safe to assume that it's not going to happen since it's a part of testcase.

Is GTEST doing that internally?
</blockquote></details>

You are correct that we won't ever ourselves call `new`, but I believe from past error experiences that the GTEST macro does not guarantee to obey Eigen alignment for our member fields.  I don't remember the detail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_q28H0VWpF2fxd0kvv:b-tja5pb) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Without it, we have the following build error:

```
external/eigen/Eigen/src/Core/GeneralProduct.h:250:58: error: value of type 'drake::symbolic::Formula' is not contextually convertible to 'bool'
      const bool alphaIsCompatible = (!ComplexByReal) || (numext::imag(actualAlpha)==RealScalar(0));
                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
</blockquote></details>

Have you tried to see if you can overload `Eigen::numext::imag` for `SymbolicExpression` to return a type other than `SymbolicExpression` that has a special one-off `operator==` with `SymbolicExpression`, but is otherwise implicitly convertible to `SymbolicExpression`?

Alternatively if would be nice if Eigen were changed to either immediately realize, or be able to traits-ask statically whether a type can ever have a non-zero imaginary component.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_q1QiplWliFwE8qRIF:b-tj1n2v) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
> GTEST macro does not guarantee to obey Eigen alignment for our member fields.

Good to know! I'll ask it to our Slack team if anyone knows more details.

So, what would be a quickest solution?
1. Add `EIGEN_MAKE_ALIGNED_OPERATOR_NEW`
2. Use `::DontAlign` template param.
3. ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_q5uzEvwwQzu6krzqh:b-hla4nv) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> GTEST macro does not guarantee to obey Eigen alignment for our member fields.

Good to know! I'll ask it to our Slack team if anyone knows more details.

So, what would be a quickest solution?
1. Add `EIGEN_MAKE_ALIGNED_OPERATOR_NEW`
2. Use `::DontAlign` template param.
3. ?
</blockquote></details>

There are only two fields here that require it (Matrix2d, Vector2d), so I like solution (2) personally.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/symbolic_expression.h, line 495 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_q0F1uAHQUoWXX5nQ2:-K_q6YMTjUBagRWZq808:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/22d4fd878e5e3d254a7bb9bc48db5fbd0e40ca38/drake/common/symbolic_expression.h#L495)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`Matrix<Expression>` instead of `MatrixBase<Expression>`?
</blockquote></details>

Done.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_q6ZRyZ965MTs3cCWZ:b-xcq2d8) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Have you tried to see if you can overload `Eigen::numext::imag` for `SymbolicExpression` to return a type other than `SymbolicExpression` that has a special one-off `operator==` with `SymbolicExpression`, but is otherwise implicitly convertible to `SymbolicExpression`?

Alternatively if would be nice if Eigen were changed to either immediately realize, or be able to traits-ask statically whether a type can ever have a non-zero imaginary component.
</blockquote></details>

I see. I'll check it this evening.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_q6f05ZmpNayPXqE5O:b-876468) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

There are only two fields here that require it (Matrix2d, Vector2d), so I like solution (2) personally.
</blockquote></details>

Good. I'll do that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_q9ycjiy1UB74i8iaP:b-ucfkn4) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I see. I'll check it this evening.
</blockquote></details>

Thanks.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_q9wx1i8P9R1jJ7PNo:b-ucfkn4) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Good. I'll do that.
</blockquote></details>

Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_qJy3RB6JrB7LaGcku:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Thanks.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/test/symbolic_expression_test.cc, line 2062 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pfc1wQoLkJbkt51Hs:-K_qKuty6lecvevW5K8U:bz864nz) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/test/symbolic_expression_test.cc#L2062)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

FYI Only the `double` ones needed DontAlign I think.  But anyway this is fine also.

---

*[drake/common/test/symbolic_variable_test.cc, line 30 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_qL6cgryER0DL2PtYx:-K_qL6chu7TaTs60ZXYc:b-y0su67) ([raw file](https://github.com/robotlocomotion/drake/blob/5e438594e74c2c8df33e8f3f46cd551b88db292c/drake/common/test/symbolic_variable_test.cc#L30)):*
> ```cpp
>   const Variable z_{"z"};
>   const Variable w_{"w"};
>   Eigen::Matrix<symbolic::Variable, 2, 2, Eigen::DontAlign> M_;
> ```

BTW This is not needed, only doubles need it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/symbolic_variable_test.cc, line 30 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_qL6cgryER0DL2PtYx:-K_qPobwP8ICGeoHQ67T:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5e438594e74c2c8df33e8f3f46cd551b88db292c/drake/common/test/symbolic_variable_test.cc#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This is not needed, only doubles need it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-K_xW_GArZFj9In0cWBE:b-an709j) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Thanks.
</blockquote></details>

I implemented it but it doesn't work yet. Here is the commit (in a separate branch): https://github.com/soonho-tri/drake/commit/7fc6aa81f796d1adf8f3044f71c9a9f82dd95f13

Interestingly, this test https://github.com/soonho-tri/drake/commit/7fc6aa81f796d1adf8f3044f71c9a9f82dd95f13#diff-52ac9ff6d0f048818f91e76a229c6e83R1080 works which I copied from the Eigen file.

Is it possible the overload `Eigen::numext::imag` that I've added is not picked up by Eigen (but used in my test code)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-Ka3Hr4AnzYZy6JhXkgu:bombirk) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I implemented it but it doesn't work yet. Here is the commit (in a separate branch): https://github.com/soonho-tri/drake/commit/7fc6aa81f796d1adf8f3044f71c9a9f82dd95f13

Interestingly, this test https://github.com/soonho-tri/drake/commit/7fc6aa81f796d1adf8f3044f71c9a9f82dd95f13#diff-52ac9ff6d0f048818f91e76a229c6e83R1080 works which I copied from the Eigen file.

Is it possible the overload `Eigen::numext::imag` that I've added is not picked up by Eigen (but used in my test code)?
</blockquote></details>

From f2f: perhaps the overload is not being used.  Our overload was `const SymbolicExpression&`), perhaps a different argument type is required.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-:-Ka3KDTTZow7mpXHUTII:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-Ka3QuDtVx08GefvuD5A:b98vebu) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

From f2f: perhaps the overload is not being used.  Our overload was `const SymbolicExpression&`), perhaps a different argument type is required.
</blockquote></details>

Adding an overloading for `SymbolicExpression&&` doesn't help either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_formula.h, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-K_pYeylg5hvJnyMJjoK:-Ka3X58TBS5EWTzkdMG5:b48xp6k) ([raw file](https://github.com/robotlocomotion/drake/blob/7a0d4a998a2cee3c2a64b4e6fdaf9bc7fa278eb2/drake/common/symbolic_formula.h#L127)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Adding an overloading for `SymbolicExpression&&` doesn't help either.
</blockquote></details>

Well, probably not worth exploring further.  Marking resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_formula.h, line 126 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-Ka3XOJi0Uu5UG3D0rqY:-Ka3XOJjae3OWm-GMsX1:bomv5v1) ([raw file](https://github.com/robotlocomotion/drake/blob/7ecdcd61ed09d91f103811b764da52de293b2e47/drake/common/symbolic_formula.h#L126)):*
> ```objc
>   static Formula False();
> 
>   /** Conversion to bool. */
> ```

Please expand this comment to provide more clarity for readers.  Something like (1) mostly only intended for third-party code doing things like `if (imag(SymbolicExpression(0)) == SymbolicExpression(0)) { ... }`; (2) prefer to call `Evaluate` from within Drake, for readability; (3) if the formula is not trivially true or false, will throw.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-:-Ka3ZRCqe1moezfg-c78:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/symbolic_formula.h, line 126 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4702#-Ka3XOJi0Uu5UG3D0rqY:-Ka3__DjInfTfzl1yXd1:b6c8we0) ([raw file](https://github.com/robotlocomotion/drake/blob/7ecdcd61ed09d91f103811b764da52de293b2e47/drake/common/symbolic_formula.h#L126)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please expand this comment to provide more clarity for readers.  Something like (1) mostly only intended for third-party code doing things like `if (imag(SymbolicExpression(0)) == SymbolicExpression(0)) { ... }`; (2) prefer to call `Evaluate` from within Drake, for readability; (3) if the formula is not trivially true or false, will throw.
</blockquote></details>

I've added a note about it at the class definition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4702)*
<!-- Sent from Reviewable.io -->
